### PR TITLE
Make blazor default for canvas; switch integration tests to F# backend

### PR DESCRIPTION
**Pull request by StachuDotNet**
#3254

## What is the problem/goal being addressed?
- Integration tests should run against the F# server (on port 9000) by default
- The canvas should use the WASM-compiled (Blazor) F# backend for analysis, rather than the OCaml one.

## What are the changes here? How do they solve the problem and what other product impacts do they cause?
*Please include before/after screenshots/gifs if appropriate*

### CODE REVIEW

**Comment by pbiggar** [comment]
If you rebase/merge this on main, CI should now run, I believe.

**Comment by StachuDotNet** [comment]
Some integration tests rely on the package manager (incl. uploading of packages).

Should this PR be blocked by that functionality, in which case I could pivot to such, or pull out into a future PR, with an FSTODO on the tests?

**Comment by pbiggar** [comment]
I would say you can do an FSTODO on it for now and we can reassess later if it's truly necessary.

**Comment by StachuDotNet** [comment]
The `test_admin/stdlib/Test::one` fn exposed [here](https://github.com/darklang/dark/blob/main/integration-tests/clear-db.sh#L43) does not pass any of the patterns that our F# backend accepts [here](https://github.com/darklang/dark/blob/main/fsharp-backend/src/LibExecution/ProgramTypes.fs#L44-L75).

image

```
Bad format in function name [2/22/2022 3:38:48 PM +00:00]
        exception: True
        exception.type: Prelude+InternalException
        exception.stacktrace: 
        exception.message: Bad format in function name
        fnName: test_admin/stdlib/Test::one
```

and so a test that uses said function (see [here](https://github.com/darklang/dark/blob/main/integration-tests/tests.ts#L1288) and [here](https://github.com/darklang/dark/blob/main/backend/test_appdata/test-package_function_references_work.json#L12)) fails when targeting the F# backend.

What should be done here @pbiggar?
- adjust the insertion somehow
- add another allowable pattern to `ProgramTypes.fs`
- adjust how the test calls/references the function
- something else

**Comment by pbiggar** [comment]
> The `test_admin/stdlib/Test::one` fn inserted [here](https://github.com/darklang/dark/blob/main/integration-tests/clear-db.sh#L43) does not pass any of the patterns that our F# backend accepts [here](https://github.com/darklang/dark/blob/main/fsharp-backend/src/LibExecution/ProgramTypes.fs#L44-L75).
> 
> image
> 
> ```
> Bad format in function name [2/22/2022 3:38:48 PM +00:00]
>         exception: True
>         exception.type: Prelude+InternalException
>         exception.stacktrace: 
>         exception.message: Bad format in function name
>         fnName: test_admin/stdlib/Test::one
> ```
> 
> and so a test that uses said function (see [here](https://github.com/darklang/dark/blob/main/integration-tests/tests.ts#L1288) and [here](https://github.com/darklang/dark/blob/main/backend/test_appdata/test-package_function_references_work.json#L12)) fails when targeting the F# backend.
> 
> What should be done here @pbiggar?
> 
> * adjust the insertion somehow
> * add another allowable pattern to `ProgramTypes.fs`
> * adjust how the test calls/references the function
> * something else

I think rename it (so adjust the insertion) so that it's valid.

File: backend/test_appdata/test-fluid_test_copy_request_as_curl.json
Comment (StachuDotNet) [inline backend/test_appdata/test-fluid_test_copy_request_as_curl.json:6]: with the original coordinates, the relevant UI was hidden under the "first time" prompts. Is this a reasonable fix?

File: backend/test_appdata/test-package_function_references_work.json
Comment (StachuDotNet) [inline backend/test_appdata/test-package_function_references_work.json:12]: Maybe something to pull out into a separate issue:
without changing this (and other corresponding locations) to be a v1 rather than a v0, the _v0 portion of the function name was stripped at the call site. Seemingly this is an OCaml interop issue, but couldn't pin down the specific issue.

File: integration-tests/README.md
Comment (StachuDotNet) [inline integration-tests/README.md:53]:
```suggestion
### Run Against OCaml

By default, the integration tests are run against the F# backend.
```

File: integration-tests/README.md
Comment (StachuDotNet) [inline integration-tests/README.md:55]:
```suggestion
If you'd like to run against the OCaml backend, use `--ocaml`
```

File: integration-tests/README.md
Comment (StachuDotNet) [inline integration-tests/README.md:155]:
```suggestion
```

File: integration-tests/run.sh
Comment (StachuDotNet) [inline integration-tests/run.sh:10]:
```suggestion
# CLEANUP this script fails if you cd to this directory.
```

File: integration-tests/tests.ts
Comment (StachuDotNet) [inline integration-tests/tests.ts:1304]: TODO for self: uncomment these lines.

File: backend/test_appdata/test-fluid_test_copy_request_as_curl.json
Comment (pbiggar) [reply backend/test_appdata/test-fluid_test_copy_request_as_curl.json:6]: The first time prompts aren't supposed to be there. We write some values to local storage to ensure they aren't so it sounds like something is broken there.

File: backend/test_appdata/test-package_function_references_work.json
Comment (pbiggar) [reply backend/test_appdata/test-package_function_references_work.json:12]: Yeah, that sounds right. This fix is fine - it's an awkward thing to fix in both versions

File: backend/test_appdata/test-fluid_test_copy_request_as_curl.json
Comment (StachuDotNet) [reply backend/test_appdata/test-fluid_test_copy_request_as_curl.json]: My issue here must have been related to something else - test is passing now with original position.

**Comment by StachuDotNet** [comment]
two steps resulted in progress on integration tests running in CI against F#:
- the `integration-tests` job was updated to run `run-fsharp-server --published` instead of `run-backend-server`
- `run-fsharp-server` itself was updated to reference the `/publish`ed binaries rather than the ones at `Release/net6.0/linux-x64/publish/`, when the `--publish` flag is present. This is to match what's persisted to workspace in the circle job
  image

Given those two changes, tests are _almost_ running in CI.
At this point, it's held up on `_wait-until-server-ready` waiting for appsupport.js to be available _on port 8000_.
image

My plan now is to update `_wait-until-server-read` so that it accepts a `--use-apiserver` flag, in which case `wait_for {}` can use `DARK_CONFIG_APISERVER_STATIC_HOST` rather than `DARK_CONFIG_STATIC_HOST` - I think that will do the trick.

File: integration-tests/tests.ts
Comment (StachuDotNet) [inline integration-tests/tests.ts:755]: TODO: this will pass in F# but not in OCaml. fix that.

File: integration-tests/tests.ts
Comment (StachuDotNet) [inline integration-tests/tests.ts:1069]: TODO: I added it but in hindsight not sure it was needed;  re-evaluate.

File: integration-tests/tests.ts
Comment (StachuDotNet) [inline integration-tests/tests.ts:1294]: Question for @pbiggar: what's the "any" for here? how does this not break?

**Comment by StachuDotNet** [comment]
Finally passing in CI!

I added a few TODOs for myself to resolve after some lunch, but ready to have another set of eyes on for now.

Some notes:
- I doubled the playwright timeout. Doesn't feel great, but acceptable?
- I'm not sure what level of detail you typically receive in your Rollbars, and if these are satisfactory.
- Relevant `docs` PR to merge as/after this merges: darklang/docs/pull#227

File: integration-tests/tests.ts
Comment (pbiggar) [reply integration-tests/tests.ts:1294]: The "Any" is the type. The text contents of the entire `div` is jammed together (even though they are in sibling sub-divs).

File: fsharp-backend/src/LibBackend/Canvas.fs
Comment (pbiggar) [inline fsharp-backend/src/LibBackend/Canvas.fs:694]: What's happening here?

File: integration-tests/tests.ts
Comment (pbiggar) [reply integration-tests/tests.ts:755]: You could pass it if it includes either the OCaml OR the F# error message.

File: integration-tests/tests.ts
Comment (pbiggar) [reply integration-tests/tests.ts:1069]: I don't understand why this would be useful. Doesn't this pause the execution? If it's needed, can we have a comment explaining it?

File: scripts/devcontainer/_wait-until-server-ready
Comment (pbiggar) [inline scripts/devcontainer/_wait-until-server-ready:10]: I think I would rather see this split out into a script called `wait-until-apiserver-ready`

**Review by pbiggar — CHANGES_REQUESTED** [review]
Looks great!

File: integration-tests/tests.ts
Comment (StachuDotNet) [reply integration-tests/tests.ts:1069]: I removed the line now and there was no change. I must have added it when diagnosing a test locally, and wanted to quickly pause at that point in debug ode.

File: integration-tests/tests.ts
Comment (StachuDotNet) [reply integration-tests/tests.ts:755]: We must have fixed this error message since I did this - the error message is consistent for me now, it seems.

File: fsharp-backend/src/LibBackend/Canvas.fs
Comment (StachuDotNet) [reply fsharp-backend/src/LibBackend/Canvas.fs:694]: At one point, I believed that "tests without a test file" were failing here. It seems they're not, though, so I can remove this try/catch now.

Comment (StachuDotNet) [reply fsharp-backend/src/LibBackend/Canvas.fs]: https://app.circleci.com/pipelines/github/darklang/dark?branch=pull%2F3485&filter=all seems I'm not crazy, actually. These failing after removing.

Comment (pbiggar) [reply fsharp-backend/src/LibBackend/Canvas.fs:694]: Ah, it looks like the old load_json_from_disk_fluid uses `File.maybeReadFile` and uses options. Do we have a preference for Options vs Exceptions? I would imagine we prefer options.

**Review by pbiggar — CHANGES_REQUESTED** [review]
One remaining question, happy to merge with or without.

File: fsharp-backend/src/LibBackend/Canvas.fs
Comment (StachuDotNet) [reply fsharp-backend/src/LibBackend/Canvas.fs]: Agreed - just pushed a PR that gets us to a slightly nicer place

**Pull request merged**
